### ISSUE TITLE
GUI callbacks not found

### ISSUE THREAD

**Issue by dusby** [secure-software-engineering/FlowDroid#178]
I maybe wrong but it seems that in Flowdroid, callbacks are looked for in lifecycle components only, is that the case ?
In fact, I have to analyze a class extending android.view.View which contains the callback method "onDraw". The interesting part of the code I have to analyze is in there but this method does not seem to be represented as a callback, therefore it does not appear in the analysis.

**Comment by dusby** [comment]
I have the same issue with all callbacks in Activities (e.g. onOptionsItemSelected, onOptionsMenuClosed, onPreparePanel, etc...), they do not appear in the analysis. I understand they are not lifecycle methods, but still it should be interesting that they appear somewhere in the component analysis.

**Comment by StevenArzt** [comment]
There is a method ```analyzeMethodOverrideCallbacks``` in class ``AbstractCallbackAnalyzer``. It should normally be invoked on every component (and therefore especially every Activity) and also on every class with a view, regardless of whether the view is registered dynamically in the code or through a layout XML file. Maybe we have a bug in that implementation.

Another reason could be a cut-off. If one component registers a view, which registers another view, and so on, we don't follow that path indefinitely. There is a limit, which is called ```maxCallbackAnalysisDepth``` in the configuration object (class ``CallbackConfiguration`` in ``InfoflowAndroidConfiguration``)

**Comment by dusby** [comment]
I do not think that ```analyzeMethodOverrideCallbacks``` method is invoked on every class with a view, look a this piece of code from ```collectCallbackMethods``` : 

```java
for (SootClass sc : entryPointClasses) {
    if (isKilled != null)
        break;
    List<MethodOrMethodContext> methods = new ArrayList<MethodOrMethodContext>(
    getLifecycleMethods(sc));
    analyzeRechableMethods(sc, methods);
    analyzeMethodOverrideCallbacks(sc);
}
```

It is called on every component (entrypoints) but not other classes like views. I have checked, the only other call to this method is in the ```collectCallbackMethods``` of ```FastCallbackAnalyzer``` class.

In the example I am working on, the main activity directly register the view which is used.

**Comment by StevenArzt** [comment]
Note that there is a second call to ``analyzeMethodOverrideCallbacks`` in the method you mentioned. Collecting callbacks is a multi-step process. First, the analyzer processes all entry points. If new classes are referenced, they are put into a worklist. Next, the worklist is processed. That's where the second call is. If this leads to nw classes, the worklist gets extended with that class as well. This process is repeated until we either have an empty worklist or the maximum depth is reached.

**Comment by dusby** [comment]
Yes you are right, I was wrong about the location of the problem. I think I found it : 

in ```findClassLayoutMappings``` method in ```DefaultCallbackAnalyzer``` class line 291 to 295 : 
```java
for (Value val : inv.getArgs()) {
    Integer intValue = valueProvider.getValue(sm, stmt, val, Integer.class);
    if (intValue != null)
        this.layoutClasses.put(sm.getDeclaringClass(), intValue);
}
```

Here, intValue is assumed to have the value that is passed as a parameter to ```setContentView(int)``` method. In my case it is null because the value is not a constant but a local as it is declared as follow : 
```
r11 = new class.implementing.View; //new View object
specialinvoke r11.<class.implementing.View: void <init>(android.content.Context)>(r12); //constructor of View object
$r0.<main.activity: class.implementing.View view1> = r11;
$r5 = $r0.<main.activity: class.implementing.View view1>; // r5 <- main.activity view1 field (class.implementing.View)
virtualinvoke $r0.<main.activity: void setContentView(android.view.View)>($r5);
```
I think this is why the class extending the android.view.View class is not analyzed for any callbacks and therefore the ```onDraw``` method is not visible.

What do you think ?

**Comment by dusby** [comment]
Ok, the issue is more subtle, look at this code : 
```java
public boolean isClassInSystemPackage(String className) {
    return (className.startsWith("android.") || className.startsWith("java.") 
        || className.startsWith("javax.")
        || className.startsWith("sun.") || className.startsWith("org.omg.")
        || className.startsWith("org.w3c.dom.") || className.startsWith("com.google.")
        || className.startsWith("com.android.")) && this.excludeSystemComponents;
}
```
Two weeks ago we added the ```excludeSystemComponents``` boolean value to take into account false system classes to be analyzed. 
The problem is here (line 224 of ```AbstractCallbackAnalyzer```) : 
```java
if (!SystemClassHandler.v().isClassInSystemPackage(iinv.getMethod().getDeclaringClass().getName()))
    continue;
```
The callback class (in this case android.location.Location, but true for others) is not take into account as a system class as the ```excludeSystemComponents``` is set to false to enable false system classes to be analyzed. So the solution is not sound.
I am thinking of a more workable solution.
